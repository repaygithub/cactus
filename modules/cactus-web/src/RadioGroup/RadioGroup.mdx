---
name: RadioGroup
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import RadioGroup from './RadioGroup'
import cactusTheme from '@repay/cactus-theme'
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live'

# RadioGroup

`RadioGroup` is a wrapper for a group of radio buttons, styled so that it fits in with and behaves like other field types (e.g. `TextInputField`).

## Basic Usage

`RadioGroup` is used like other `AccessibleField` derivatives, except that the individual radio buttons are passed as children. The `name`, `required`, and `disabled` props are passed to the `RadioGroup` and forwarded to the individual buttons; the first three are unconditional, but `disabled` is only forwarded if it is `true` (so that you can disable individual buttons if needed).

`RadioGroup.Button` is a straight wrapper around `RadioButtonField`, with the `name`, and `required` props removed so they can be omitted without causing Typescript errors.

### Try it out

export const code = `<RadioGroup name="radioCity" label="Here there be radios" value="option_b">
  <RadioGroup.Button label="Radio The First" value="option_a" />
  <RadioGroup.Button label="Radio The Second" value="option_b" />
  <RadioGroup.Button label="Radio The Third" value="option_c" />
</RadioGroup>`

<LiveProvider code={code} scope={{ RadioGroup }}>
  <LiveEditor style={livePreviewStyle} />
  <LiveError />
  <LivePreview />
</LiveProvider>

## Best practices

- Use meaningful labels to ensure accessibility.
- Prop forwarding for `name`, etc., will only work if the buttons are direct children of the group (or inside a `React.Fragment`, array, etc.). If you use an intermediate component, it will need to accept the forwarded props and pass them on in turn.
- You can control the selected value in one of two ways (which should not be mixed):
  - You can pass `value` or `defaultValue` to the group.
  - You can pass `checked` or `defaultChecked` to individual buttons.
  - This also works by prop forwarding, and has a similar caveat: if the child has a `value` prop that matches the group's `value` prop, a `checked` prop is added (and similarly for `defaultValue`/`defaultChecked`); **if the child has no `value` prop the group's `value` is forwarded,** so keep that in mind if you need a radio button with no value or have an intermediate component.
- A `ref` passed to `RadioGroup` will be set on the `<fieldset>` HTML element that wraps the group.
- The `onChange`, `onFocus`, and `onBlur` handlers are _not_ forwarded to the individual buttons; they are attached to the `<fieldset>` and rely on event delegation to capture changes.
  - This especially impacts the focus handlers, because when switching from one button to the next within the group, you'll get a blur event for the first button followed by a focus event for the second.

## Properties

<PropsTable of={RadioGroup} includeProps={['name', 'label', 'disabled', 'tooltip', 'error', 'warning', 'success']} />
