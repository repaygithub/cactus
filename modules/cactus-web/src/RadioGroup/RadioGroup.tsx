import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import { margin, MarginProps, width, WidthProps } from 'styled-system'

import { FieldProps, useAccessibleField } from '../AccessibleField/AccessibleField'
import Box from '../Box/Box'
import FieldWrapper from '../FieldWrapper/FieldWrapper'
import handleEvent from '../helpers/eventHandler'
import { cloneAll, useMergedRefs } from '../helpers/react'
import { border } from '../helpers/theme'
import Label from '../Label/Label'
import RadioButtonField, { RadioButtonFieldProps } from '../RadioButtonField/RadioButtonField'
import StatusMessage from '../StatusMessage/StatusMessage'
import Tooltip from '../Tooltip/Tooltip'
import { FieldOnBlurHandler, FieldOnChangeHandler, FieldOnFocusHandler } from '../types'

interface RadioGroupProps
  extends MarginProps,
    WidthProps,
    Omit<FieldProps, 'labelProps'>,
    Omit<
      React.FieldsetHTMLAttributes<HTMLFieldSetElement>,
      'name' | 'onChange' | 'onFocus' | 'onBlur'
    > {
  value?: string | number
  defaultValue?: string | number
  required?: boolean
  onChange?: FieldOnChangeHandler<string>
  onFocus?: FieldOnFocusHandler
  onBlur?: FieldOnBlurHandler
}

// This allows omitting the required `name` prop, since it'll be injected by the RadioGroup.
type RadioGroupButtonProps = Omit<RadioButtonFieldProps, 'name' | 'required'>
const RadioGroupButton = React.forwardRef<HTMLInputElement, RadioGroupButtonProps>(
  (props: any, ref) => <RadioButtonField ref={ref} {...props} />
)
RadioGroupButton.displayName = 'RadioGroup.Button'

type ForwardProps = {
  name: string
  disabled?: boolean
  required?: boolean
}

const useValue = (
  fsRef: React.RefObject<HTMLFieldSetElement>,
  attr: 'checked' | 'defaultChecked',
  value: string | number | undefined,
  dependencies: any[] = []
) => {
  React.useEffect(() => {
    if (value !== undefined && fsRef.current instanceof HTMLFieldSetElement) {
      const radios = fsRef.current.querySelectorAll('input[type="radio"]')
      for (let i = 0; i < radios.length; i++) {
        const radio = radios[i] as HTMLInputElement
        if (radio.value == value) {
          radio[attr] = true
        }
      }
    }
  }, dependencies) // eslint-disable-line react-hooks/exhaustive-deps
}

export const RadioGroup = React.forwardRef<HTMLFieldSetElement, RadioGroupProps>(
  (
    {
      label,
      children,
      tooltip,
      required,
      defaultValue,
      value,
      onChange,
      onFocus,
      onBlur,
      ...props
    },
    ref
  ) => {
    const {
      fieldId,
      ariaDescribedBy,
      labelId,
      statusId,
      name,
      status,
      statusMessage,
      tooltipId,
      disabled,
    } = useAccessibleField(props)
    const [showTooltip, setTooltipVisible] = React.useState<boolean>(false)

    const fsRef = React.useRef<HTMLFieldSetElement>(null)
    const mergedRef = useMergedRefs(ref, fsRef)
    useValue(fsRef, 'defaultChecked', defaultValue)
    useValue(fsRef, 'checked', value, [value])

    const forwardProps: ForwardProps = { name, required }
    if (disabled === true) {
      forwardProps.disabled = disabled
    }
    children = cloneAll(children, forwardProps)

    const handleChange = React.useCallback(
      (event: React.FormEvent<HTMLFieldSetElement>): void => {
        if (typeof onChange === 'function') {
          const target = (event.target as unknown) as HTMLInputElement
          onChange(name, target.value)
        }
      },
      [name, onChange]
    )
    const handleFocus = React.useCallback(
      (e: React.FocusEvent<HTMLFieldSetElement>) => {
        if (!e.currentTarget.contains(e.relatedTarget as Node)) {
          handleEvent(onFocus, name)
          setTooltipVisible(() => true)
        }
      },
      [onFocus, name, setTooltipVisible]
    )
    const handleBlur = React.useCallback(
      (e: React.FocusEvent<HTMLFieldSetElement>) => {
        if (!e.currentTarget.contains(e.relatedTarget as Node)) {
          handleEvent(onBlur, name)
          setTooltipVisible(() => false)
        }
      },
      [onBlur, name, setTooltipVisible]
    )

    return (
      <Fieldset
        {...props}
        ref={mergedRef}
        id={fieldId}
        role="radiogroup"
        aria-describedby={ariaDescribedBy}
        name={name}
        disabled={disabled}
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
      >
        <Label as="legend" id={labelId}>
          {label}
        </Label>
        <Box mx={4} pt={3}>
          {children}
        </Box>
        {tooltip && (
          <Tooltip id={tooltipId} label={tooltip} disabled={disabled} forceVisible={showTooltip} />
        )}
        {status && (
          <div>
            <StatusMessage status={status} id={statusId}>
              {statusMessage}
            </StatusMessage>
          </div>
        )}
      </Fieldset>
    )
  }
)

RadioGroup.displayName = 'RadioGroup'
RadioGroup.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  label: PropTypes.node.isRequired,
  tooltip: PropTypes.node,
  success: PropTypes.node,
  warning: PropTypes.node,
  error: PropTypes.node,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  value: function (props: Record<string, any>): Error | null {
    if (props.value !== undefined) {
      const proptype = typeof props.value
      if (!(proptype === 'string' || proptype === 'number')) {
        return new Error('The `value` prop must be a string or number.')
      }
    }
    return null
  },
  defaultValue: function (props: Record<string, any>): Error | null {
    if (props.defaultValue !== undefined) {
      const proptype = typeof props.defaultValue
      if (props.value !== undefined) {
        return new Error(
          'You provided both `value` and `defaultValue` props. Radio groups must be either controlled or uncontrolled (specify either the `value` prop, or the `defaultValue` prop, but not both). More info: https://fb.me/react-controlled-components'
        )
      } else if (!(proptype === 'string' || proptype === 'number')) {
        return new Error('The `defaultValue` prop must be a string or number.')
      }
    }
    return null
  },
}

type RadioGroupType = typeof RadioGroup & { Button: typeof RadioGroupButton }

const DefaultRadioGroup = RadioGroup as any
DefaultRadioGroup.Button = RadioGroupButton

export default DefaultRadioGroup as RadioGroupType

// The `legend` is floated to get it to position the border correctly.
const Fieldset = styled(FieldWrapper)<{ disabled?: boolean }>`
  position: relative;
  border: 0;
  margin: 0;
  padding: 0;
  ${margin}
  ${width}

  legend {
    box-sizing: border-box;
    border-bottom: ${(p) => border(p.theme, 'currentcolor')};
    padding-left: 16px;
    padding-right: 28px;
    width: 100%;
    float: left;
    + * {
      clear: both;
    }
    color: ${(p) => (p.disabled ? p.theme.colors.mediumGray : 'currentcolor')};
  }

  ${Tooltip} {
    position: absolute;
    right: 8px;
    top: 2px;
    font-size: 16px;
  }

  ${StatusMessage} {
    margin-top: 4px;
  }
`.withComponent('fieldset')
