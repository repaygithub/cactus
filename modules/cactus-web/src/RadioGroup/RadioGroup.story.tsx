import { boolean, select, text } from '@storybook/addon-knobs'
import { Meta } from '@storybook/react/types-6-0'
import React from 'react'

import Flex from '../Flex/Flex'
import RadioGroup from './RadioGroup'

export default {
  title: 'RadioGroup',
  component: RadioGroup,
} as Meta
export const BasicUsage = (): React.ReactElement => (
  <RadioGroup
    id="my-id"
    name="you-are-group-one"
    label={text('label', 'A Label')}
    disabled={boolean('disabled', false)}
    onChange={(e: any) => console.log(`'${e.target.name}' changed: ${e.target.value}`)}
    onFocus={(e: any) => console.log(`'${e.target.value}' focused`)}
    onBlur={(e: any) => console.log(`'${e.target.value}' blurred`)}
    tooltip={text('tooltip', 'Here there be radio buttons')}
    error={text('error', '')}
    success={text('success', '')}
    warning={text('warning', '')}
    autoTooltip={boolean('autoTooltip', true)}
    disableTooltip={select('disableTooltip', [false, true, undefined], false)}
  >
    <RadioGroup.Button label="That's right" value="right" />
    <RadioGroup.Button disabled label="That's wrong" value="left" />
    <RadioGroup.Button label={text('button label', "That's...")} value="center" />
  </RadioGroup>
)

export const WithValues = (): React.ReactElement => {
  const [value, setValue] = React.useState<string>('strong')
  return (
    <Flex>
      <RadioGroup
        name="youAreGroupTwo"
        label="Controller"
        value={value}
        onChange={(e: any) => setValue(e.target.value)}
      >
        <RadioGroup.Button label="Empty" value="" />
        <RadioGroup.Button label={<sup>Strong</sup>} value="strong" />
        <RadioGroup.Button label="Fortunate" value="ladybug" />
      </RadioGroup>
      <RadioGroup name="youAreGroupThree" label={<em>Follower</em>} required>
        <RadioGroup.Button checked={value === ''} label="Zen" value="" />
        <RadioGroup.Button checked={value === 'strong'} label="Confident" value="strong" />
        <RadioGroup.Button checked={value === 'ladybug'} label="Miraculous" value="ladybug" />
      </RadioGroup>
    </Flex>
  )
}

export const WithDefaultValues = (): React.ReactElement => {
  const group = React.useRef<HTMLFormElement>(null)
  const button = React.useRef<HTMLFormElement>(null)
  return (
    <Flex>
      <form ref={group}>
        <RadioGroup name="youAreGroupFour" label="On Group" defaultValue="punny">
          <RadioGroup.Button label="Pining" value="sigh" />
          <RadioGroup.Button label="UnFortunate" value="catNoir" />
          <RadioGroup.Button label="Freedom" value="punny" />
        </RadioGroup>
        <button type="button" onClick={() => group.current?.reset()}>
          Reset
        </button>
      </form>
      <form ref={button}>
        <RadioGroup name="youAreGroupFive" label="On Button">
          <RadioGroup.Button label="Alone" value="sigh" />
          <RadioGroup.Button label="Destruction" value="catNoir" />
          <RadioGroup.Button defaultChecked label="Unleashed" value="punny" />
        </RadioGroup>
        <button type="button" onClick={() => button.current?.reset()}>
          Reset
        </button>
      </form>
    </Flex>
  )
}
