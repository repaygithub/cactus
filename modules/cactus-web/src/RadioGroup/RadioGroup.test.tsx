import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as React from 'react'

import { StyleProvider } from '../StyleProvider/StyleProvider'
import RadioGroup from './RadioGroup'

describe('component: RadioGroup', (): void => {
  test('should render a radio group', (): void => {
    const { container, getByText } = render(
      <StyleProvider>
        <RadioGroup
          id="rg"
          name="places"
          label="Realm"
          required
          disabled={false}
          value="persephone"
          tooltip="Select your preferred realm"
          onChange={() => undefined}
        >
          <RadioGroup.Button id="earth" label="Glebe" value="james" />
          <RadioGroup.Button id="life" label="Elysium" value="persephone" />
          <RadioGroup.Button id="air" label="Empyrea" value="stratos" />
          <RadioGroup.Button id="fire" label="Pyroborea" value="pyro" />
          <RadioGroup.Button id="death" label="Stygia" value="charnel" disabled />
        </RadioGroup>
      </StyleProvider>
    )

    const radios = container.querySelectorAll('input')
    for (let i = 0; i < 5; i++) {
      const radio = radios[i]
      expect(radio.name).toBe('places')
      expect(radio.type).toBe('radio')
      expect(radio.required).toBe(true)
      expect(radio.disabled).toBe(i === 4)
      expect(radio.checked).toBe(i === 1)
    }

    const tooltip = getByText('Select your preferred realm')
    expect(tooltip).toHaveAttribute('id', 'rg-tip')
    expect(tooltip).toHaveAttribute('role', 'tooltip')
  })

  test('should render a disabled radio group', (): void => {
    const { container, getByText } = render(
      <StyleProvider>
        <RadioGroup
          id="rg"
          name="cities"
          label="Captitol"
          disabled
          defaultValue="stratos"
          warning="Cannot travel right now"
        >
          <RadioGroup.Button id="earth" label="Agothera" value="james" />
          <RadioGroup.Button id="life" label="Idylliac" value="persephone" />
          <RadioGroup.Button id="air" label="Thryhring" value="stratos" />
          <RadioGroup.Button id="fire" label="Helios" value="pyro" disabled={false} />
          <RadioGroup.Button id="death" label="Dys" value="charnel" />
        </RadioGroup>
      </StyleProvider>
    )

    const radios = container.querySelectorAll('input')
    for (let i = 0; i < 5; i++) {
      const radio = radios[i]
      expect(radio.name).toBe('cities')
      expect(radio.required).toBe(false)
      expect(radio.disabled).toBe(radio.value !== 'pyro')
      expect(radio.checked).toBe(i === 2)
    }

    const warning = getByText('Cannot travel right now').closest('div')
    expect(warning).toHaveAttribute('id', 'rg-status')
    expect(warning).toHaveAttribute('role', 'alert')
  })

  test('should trigger events', (): void => {
    const changes: [string, string][] = []
    const onChange = jest.fn((e) => changes.push([e.target.name, e.target.value]))
    const changeOne: [string, string][] = []
    const onChangeOne = jest.fn((e) => changeOne.push([e.target.name, e.target.value]))
    const focusNames: string[] = []
    const onFocus = jest.fn((e) => focusNames.push(e.target.id))
    const focusOne: string[] = []
    const onFocusOne = jest.fn((e) => focusOne.push(e.target.id))
    const blurNames: string[] = []
    const onBlur = jest.fn((e) => blurNames.push(e.target.id))
    const { getByLabelText } = render(
      <StyleProvider>
        <RadioGroup
          id="rg"
          name="wizards"
          label="Allies"
          onChange={onChange}
          onFocus={onFocus}
          onBlur={onBlur}
        >
          <RadioGroup.Button id="earth" label="Grakkus" value="james" onFocus={onFocusOne} />
          <RadioGroup.Button id="life" label="Yogo" value="persephone" />
          <RadioGroup.Button id="air" label="Jadugarr" value="stratos" disabled />
          <RadioGroup.Button id="fire" label="Sorcha" value="pyro" onChange={onChangeOne} />
          <RadioGroup.Button id="death" label="Acheron" value="charnel" />
        </RadioGroup>
      </StyleProvider>
    )

    userEvent.tab()
    expect(onChange).not.toHaveBeenCalled()
    expect(onChangeOne).not.toHaveBeenCalled()
    expect(onFocus).toHaveBeenCalledTimes(1)
    expect(onFocusOne).toHaveBeenCalledTimes(1)
    expect(onBlur).not.toHaveBeenCalled()

    userEvent.click(getByLabelText('Jadugarr'))
    expect(onChange).not.toHaveBeenCalled()
    expect(onChangeOne).not.toHaveBeenCalled()
    expect(onFocus).toHaveBeenCalledTimes(1)
    expect(onFocusOne).toHaveBeenCalledTimes(1)
    expect(onBlur).not.toHaveBeenCalled()

    userEvent.click(getByLabelText('Sorcha'))
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChangeOne).toHaveBeenCalledTimes(1)
    expect(onFocus).toHaveBeenCalledTimes(2)
    expect(onFocusOne).toHaveBeenCalledTimes(1)
    expect(onBlur).toHaveBeenCalledTimes(1)

    userEvent.click(getByLabelText('Yogo'))
    expect(onChange).toHaveBeenCalledTimes(2)
    expect(onChangeOne).toHaveBeenCalledTimes(1)
    expect(onFocus).toHaveBeenCalledTimes(3)
    expect(onFocusOne).toHaveBeenCalledTimes(1)
    expect(onBlur).toHaveBeenCalledTimes(2)

    userEvent.tab()
    expect(changes).toEqual([
      ['wizards', 'pyro'],
      ['wizards', 'persephone'],
    ])
    expect(changeOne).toEqual([['wizards', 'pyro']])
    expect(focusNames).toEqual(['earth', 'fire', 'life'])
    expect(focusOne).toEqual(['earth'])
    expect(blurNames).toEqual(['earth', 'fire', 'life'])
  })
})
