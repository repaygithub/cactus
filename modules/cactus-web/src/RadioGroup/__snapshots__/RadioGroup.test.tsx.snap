// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: RadioGroup should render a disabled radio group 1`] = `
.c5 + .c4 {
  margin-top: 16px;
}

.c2 {
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(200,10%,20%);
}

.c14 {
  vertical-align: middle;
}

.c12 {
  padding: 2px 4px;
  position: relative;
  box-sizing: border-box;
  overflow-wrap: break-word;
  display: inline-block;
  font-size: 15px;
  line-height: 1.6;
  background-color: hsla(47,82%,47%,0.3);
}

.c12 .sc-fzoPby,
.c12 .c13,
.c12 .sc-paXsP {
  margin-right: 4px;
  vertical-align: -2px;
}

.c3 {
  box-sizing: border-box;
  margin-left: 16px;
  margin-right: 16px;
  padding-top: 8px;
}

.c9 {
  opacity: 0;
  height: 0;
  width: 0;
  margin: 0;
  position: absolute;
}

.c10 {
  display: inline-block;
  box-sizing: border-box;
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background-color: hsl(0,0%,90%);
  border: 2px solid hsl(0,0%,90%);
}

.c10::after {
  position: absolute;
  content: '';
  display: none;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  background-color: hsl(0,0%,90%);
  margin: 2px 2px;
  box-sizing: border-box;
}

.c8 {
  position: relative;
  display: inline-block;
  vertical-align: -1px;
  line-height: 16px;
  width: 16px;
  height: 16px;
  cursor: cursor;
}

.c8 input:checked ~ span {
  background-color: transparent;
}

.c8 input:checked ~ span::after {
  display: block;
}

.c8 input:focus ~ span {
  box-shadow: 0px 3px 8px hsla(200,96%,35%,0.3);
}

.c7 + .c6 {
  margin-top: 8px;
}

.c7 .c1 {
  cursor: not-allowed;
  padding-left: 8px;
  color: hsl(0,0%,70%);
}

.c0 {
  position: relative;
  border: 0;
  margin: 0;
  padding: 0;
}

.c0 legend {
  box-sizing: border-box;
  border-bottom: 1px solid currentcolor;
  padding-left: 16px;
  padding-right: 28px;
  width: 100%;
  float: left;
  color: hsl(0,0%,70%);
}

.c0 legend + * {
  clear: both;
}

.c0 .sc-pReKu {
  position: absolute;
  right: 8px;
  top: 2px;
  font-size: 16px;
}

.c0 .c11 {
  margin-top: 4px;
}

@media screen and (min-width:1024px) {
  .c2 {
    font-size: 18px;
    line-height: 1.5;
  }
}

@media screen and (min-width:1024px) {
  .c12 {
    font-size: 15px;
    line-height: 1.6;
  }
}

<div>
  <fieldset
    aria-describedby="rg-tip rg-status"
    class="c0"
    disabled=""
    id="rg"
    name="cities"
    role="radiogroup"
  >
    <legend
      class="c1 c2"
      id="rg-label"
    >
      Captitol
    </legend>
    <div
      class="c3"
    >
      <div
        class="c4 c5 c6 c7"
      >
        <label
          class="c8"
          for="earth"
        >
          <input
            class="c9"
            disabled=""
            id="earth"
            name="cities"
            type="radio"
            value="james"
          />
          <span
            aria-hidden="true"
            class="c10"
            disabled=""
          />
        </label>
        <label
          class="c1 c2"
          for="earth"
        >
          Agothera
        </label>
      </div>
      <div
        class="c4 c5 c6 c7"
      >
        <label
          class="c8"
          for="life"
        >
          <input
            class="c9"
            disabled=""
            id="life"
            name="cities"
            type="radio"
            value="persephone"
          />
          <span
            aria-hidden="true"
            class="c10"
            disabled=""
          />
        </label>
        <label
          class="c1 c2"
          for="life"
        >
          Idylliac
        </label>
      </div>
      <div
        class="c4 c5 c6 c7"
      >
        <label
          class="c8"
          for="air"
        >
          <input
            checked=""
            class="c9"
            disabled=""
            id="air"
            name="cities"
            type="radio"
            value="stratos"
          />
          <span
            aria-hidden="true"
            class="c10"
            disabled=""
          />
        </label>
        <label
          class="c1 c2"
          for="air"
        >
          Thryhring
        </label>
      </div>
      <div
        class="c4 c5 c6 c7"
      >
        <label
          class="c8"
          for="fire"
        >
          <input
            class="c9"
            disabled=""
            id="fire"
            name="cities"
            type="radio"
            value="pyro"
          />
          <span
            aria-hidden="true"
            class="c10"
            disabled=""
          />
        </label>
        <label
          class="c1 c2"
          for="fire"
        >
          Helios
        </label>
      </div>
      <div
        class="c4 c5 c6 c7"
      >
        <label
          class="c8"
          for="death"
        >
          <input
            class="c9"
            disabled=""
            id="death"
            name="cities"
            type="radio"
            value="charnel"
          />
          <span
            aria-hidden="true"
            class="c10"
            disabled=""
          />
        </label>
        <label
          class="c1 c2"
          for="death"
        >
          Dys
        </label>
      </div>
    </div>
    <div>
      <div
        class="c11 c12"
        id="rg-status"
        role="alert"
      >
        <svg
          aria-hidden="true"
          class="c13 c14"
          fill="currentcolor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M1.39339828,12.7071068 C1.00287399,12.3165825 1.00287399,11.6834175 1.39339828,11.2928932 L11.2928932,1.39339828 C11.6834175,1.00287399 12.3165825,1.00287399 12.7071068,1.39339828 L22.6066017,11.2928932 C22.997126,11.6834175 22.997126,12.3165825 22.6066017,12.7071068 L12.7071068,22.6066017 C12.3165825,22.997126 11.6834175,22.997126 11.2928932,22.6066017 L1.39339828,12.7071068 Z M3.51471863,12 L12,20.4852814 L20.4852814,12 L12,3.51471863 L3.51471863,12 Z M12,16 C11.4477153,16 11,15.5522847 11,15 C11,14.4477153 11.4477153,14 12,14 C12.5522847,14 13,14.4477153 13,15 C13,15.5522847 12.5522847,16 12,16 Z M13,12 C13,12.5522847 12.5522847,13 12,13 C11.4477153,13 11,12.5522847 11,12 L11,9 C11,8.44771525 11.4477153,8 12,8 C12.5522847,8 13,8.44771525 13,9 L13,12 Z"
          />
        </svg>
        <span>
          Cannot travel right now
        </span>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`component: RadioGroup should render a radio group 1`] = `
.c5 + .c4 {
  margin-top: 16px;
}

.c2 {
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(200,10%,20%);
}

.c15 {
  vertical-align: middle;
}

.c16 {
  outline: none;
  color: hsl(200,10%,20%);
}

.c16:hover {
  color: hsl(200,96%,35%);
}

.c3 {
  box-sizing: border-box;
  margin-left: 16px;
  margin-right: 16px;
  padding-top: 8px;
}

.c9 {
  opacity: 0;
  height: 0;
  width: 0;
  margin: 0;
  position: absolute;
}

.c10 {
  display: inline-block;
  box-sizing: border-box;
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background-color: transparent;
  border: 2px solid hsl(200,96%,11%);
}

.c10::after {
  position: absolute;
  content: '';
  display: none;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  background-color: hsl(200,96%,35%);
  margin: 2px 2px;
  box-sizing: border-box;
}

.c13 {
  display: inline-block;
  box-sizing: border-box;
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background-color: hsl(0,0%,90%);
  border: 2px solid hsl(0,0%,90%);
}

.c13::after {
  position: absolute;
  content: '';
  display: none;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  background-color: hsl(0,0%,90%);
  margin: 2px 2px;
  box-sizing: border-box;
}

.c8 {
  position: relative;
  display: inline-block;
  vertical-align: -1px;
  line-height: 16px;
  width: 16px;
  height: 16px;
  cursor: pointer;
}

.c8 input:checked ~ span {
  border-color: hsl(200,96%,35%);
  background-color: transparent;
}

.c8 input:checked ~ span::after {
  display: block;
}

.c8 input:focus ~ span {
  box-shadow: 0px 3px 8px hsla(200,96%,35%,0.3);
}

.c12 {
  position: relative;
  display: inline-block;
  vertical-align: -1px;
  line-height: 16px;
  width: 16px;
  height: 16px;
  cursor: cursor;
}

.c12 input:checked ~ span {
  background-color: transparent;
}

.c12 input:checked ~ span::after {
  display: block;
}

.c12 input:focus ~ span {
  box-shadow: 0px 3px 8px hsla(200,96%,35%,0.3);
}

.c7 + .c6 {
  margin-top: 8px;
}

.c7 .c1 {
  cursor: pointer;
  padding-left: 8px;
}

.c11 + .c6 {
  margin-top: 8px;
}

.c11 .c1 {
  cursor: not-allowed;
  padding-left: 8px;
  color: hsl(0,0%,70%);
}

.c0 {
  position: relative;
  border: 0;
  margin: 0;
  padding: 0;
}

.c0 legend {
  box-sizing: border-box;
  border-bottom: 1px solid currentcolor;
  padding-left: 16px;
  padding-right: 28px;
  width: 100%;
  float: left;
  color: currentcolor;
}

.c0 legend + * {
  clear: both;
}

.c0 .c14 {
  position: absolute;
  right: 8px;
  top: 2px;
  font-size: 16px;
}

.c0 .sc-qYSYK {
  margin-top: 4px;
}

@media screen and (min-width:1024px) {
  .c2 {
    font-size: 18px;
    line-height: 1.5;
  }
}

<div>
  <fieldset
    aria-describedby="rg-tip rg-status"
    class="c0"
    id="rg"
    name="places"
    role="radiogroup"
  >
    <legend
      class="c1 c2"
      id="rg-label"
    >
      Realm
    </legend>
    <div
      class="c3"
    >
      <div
        class="c4 c5 c6 c7"
      >
        <label
          class="c8"
          for="earth"
        >
          <input
            class="c9"
            id="earth"
            name="places"
            required=""
            type="radio"
            value="james"
          />
          <span
            aria-hidden="true"
            class="c10"
          />
        </label>
        <label
          class="c1 c2"
          for="earth"
        >
          Glebe
        </label>
      </div>
      <div
        class="c4 c5 c6 c7"
      >
        <label
          class="c8"
          for="life"
        >
          <input
            checked=""
            class="c9"
            id="life"
            name="places"
            required=""
            type="radio"
            value="persephone"
          />
          <span
            aria-hidden="true"
            class="c10"
          />
        </label>
        <label
          class="c1 c2"
          for="life"
        >
          Elysium
        </label>
      </div>
      <div
        class="c4 c5 c6 c7"
      >
        <label
          class="c8"
          for="air"
        >
          <input
            class="c9"
            id="air"
            name="places"
            required=""
            type="radio"
            value="stratos"
          />
          <span
            aria-hidden="true"
            class="c10"
          />
        </label>
        <label
          class="c1 c2"
          for="air"
        >
          Empyrea
        </label>
      </div>
      <div
        class="c4 c5 c6 c7"
      >
        <label
          class="c8"
          for="fire"
        >
          <input
            class="c9"
            id="fire"
            name="places"
            required=""
            type="radio"
            value="pyro"
          />
          <span
            aria-hidden="true"
            class="c10"
          />
        </label>
        <label
          class="c1 c2"
          for="fire"
        >
          Pyroborea
        </label>
      </div>
      <div
        class="c4 c5 c6 c11"
      >
        <label
          class="c12"
          for="death"
        >
          <input
            class="c9"
            disabled=""
            id="death"
            name="places"
            required=""
            type="radio"
            value="charnel"
          />
          <span
            aria-hidden="true"
            class="c13"
            disabled=""
          />
        </label>
        <label
          class="c1 c2"
          for="death"
        >
          Stygia
        </label>
      </div>
    </div>
    <span
      class="c14 "
      data-reach-tooltip-trigger=""
    >
      <svg
        class="c15 c16"
        fill="currentcolor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M22,12 C22,6.47714622 17.5223159,2 12,2 C6.47523326,2 2,6.47577095 2,12 C2,17.524229 6.47523326,22 12,22 C17.5223159,22 22,17.5228538 22,12 Z M24,12 C24,18.6274771 18.6268316,24 12,24 C5.37060996,24 0,18.6287448 0,12 C0,5.37125524 5.37060996,0 12,0 C18.6268316,0 24,5.37252293 24,12 Z M11.057,15.6442233 L11.057,12.5 L9.27493729,12.5 C8.72265254,12.5 8.27493729,12.0522847 8.27493729,11.5 C8.27493729,10.9477153 8.72265254,10.5 9.27493729,10.5 L12.057,10.5 C12.6092847,10.5 13.057,10.9477153 13.057,11.5 L13.057,15.6442233 L15.0505322,15.6442233 C15.602817,15.6442233 16.0505322,16.0919386 16.0505322,16.6442233 C16.0505322,17.1965081 15.602817,17.6442233 15.0505322,17.6442233 L9,17.6442233 C8.44771525,17.6442233 8,17.1965081 8,16.6442233 C8,16.0919386 8.44771525,15.6442233 9,15.6442233 L11.057,15.6442233 Z M11.5,9 C10.6715729,9 10,8.32842712 10,7.5 C10,6.67157288 10.6715729,6 11.5,6 C12.3284271,6 13,6.67157288 13,7.5 C13,8.32842712 12.3284271,9 11.5,9 Z"
        />
      </svg>
    </span>
    <span
      id="rg-tip"
      role="tooltip"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap; word-wrap: normal;"
    >
      Select your preferred realm
    </span>
  </fieldset>
</div>
`;
